tobler/util/raster_tools.py

```diff
@@ -0,0 +1,63 @@
+"""tools for dealing with rasters."""
+
+import ast
+
+import geopandas as gpd
+import numpy as np
+import pandas as pd
+import rasterio as rio
+from rasterio import features
+from shapely.geometry import shape
+
+
+def _parse_geom(geom_str):
+    return shape(ast.literal_eval(geom_str))
+
+
+def extract_raster_features(gdf, raster_path, pixel_types=None, nodata=255):
+    """Generate a geodataframe from raster pixels.
+
+    Parameters
+    ----------
+    gdf : geopandas.GeoDataFrame
+        geodataframe  defining the area of interest. The input raster will be
+        clipped to the extent of the geodataframe
+    raster_path : str
+        path to raster file, such as downloaded from <https://lcviewer.vito.be/download>
+    pixel_types : list-like, optional
+        subset of pixel values to extract, by default None. If None, this function
+        may generate a very large geodataframe
+    nodata : int, optional
+        pixel value denoting "no data" in input raster
+
+    Returns
+    -------
+    geopandas.GeoDataFrame
+        geodataframe whose rows are the zones extracted by the rasterio.features module.
+        The geometry of each zone is the boundary of a contiguous group of pixels with 
+        the same value; the `value` column contains the pixel value of each zone.
+    """
+    with rio.open(raster_path) as src:  # read nationwide file
+
+        raster_crs = src.crs.data
+        gdf = gdf.to_crs(raster_crs)
+        geomask = [gdf.explode().unary_union.__geo_interface__]
+
+        out_image, out_transform = rio.mask.mask(
+            src, geomask, nodata=nodata, crop=True
+        )  # clip to AoI using a vector layer
+
+        if pixel_types:
+            mask = np.isin(out_image, pixel_types)  # only include developed pixels
+            shapes = list(
+                features.shapes(out_image, mask=mask, transform=out_transform)
+            )  # convert regions to polygons
+        else:
+            shapes = list(features.shapes(out_image, transform=out_transform))
+    res = list(zip(*shapes))
+    geoms = pd.Series(res[0], name="geometry").apply(str).apply(_parse_geom)
```

Comment:
At the risk of sounding repetitive, this is a very obvious candidate to ship in chunks to different cores. A simple `joblib` set up should be able to do it. Split `res[0]` in `cpu_cores`, ship them there and reassemble into a single `Series`

Reply:
i knew it was coming--and that's part of why i opened this PR with the initial implementation, so you'd suggest the easiest path forward :P

Reply:
This may be even easier with `dask.dataframe`. Someting along this should do it.
```py
import dask.dataframe as dd
from dask.system import cpu_count

ds = dd.from_array(res[0], chunksize=len(res[0]) // cpu_count() + 1)
geoms = (ds.astype(str).apply( _parse_geom, meta=(None, "object")).compute()

```